### ISSUE TITLE
Update: Throughout - Index Operation Semantics

### ISSUE THREAD

**Issue by wesbland** [mpi-forum/mpi-issues#237]
## Problem

Index all (relevant) references to semantics of operations.  See, e.g., page 523.

### Suggested Fix

Edit to update

### Instructions

* Each committee chair should check the box when they're done. 
* All changes should be grouped into single PR which will be reviewed all at once.
  * The first author of the PR will create a new PR. All other authors will push to the first PR (instructions to be added when the first PR is created).
  * This PR should say that it `Fixes #237` in the description to automatically link it to this issue.

If there is nothing to do in your chapter, just check your box.

* [x] Front matter - @wgropp 
* [x] Introduction - @wgropp 
* [x] Terms - @cblaas 
* [ ] #611
* [x] Datatypes - @bosilca 
* [x] Collectives - @tonyskjellum 
* [x] Communicators - @GuillaumeMercier 
* [x] Topologies - @RolfRabenseifner 
* [x] Partitioned Communication - @regrant 
* [x] Environment - @bosilca 
* [x] Info - @schulzm 
* [x] Process Creation - @hppritcha 
* [ ] #612
* [x] External Interfaces - @schulzm 
* [x] I/O - @tonyskjellum 
* [ ] #613
* [x] Deprecated - @RolfRabenseifner 
* [x] Removed - @RolfRabenseifner 
* [x] Backward Incompatibilities - @wesbland 
* [ ] #614
* [x] Bindings Summary - @RolfRabenseifner 
* [x] Change Log - @RolfRabenseifner

**Comment by wesbland** [comment]
This is probably another thing that requires each chapter committee to pitch in. @wgropp can you give a more concrete example here for us to use?

* Editor - 3
* Chapter committee - 4
* Errata - 0
* Full Proposal - 0

**Comment by wgropp** [comment]
There are a lot of places where semantics are mentioned in the text, but there isn't a clear match to an entry in the index. For example, the Tools interface mentions semantics, but there are no corresponding index entries.  Also, under semantics in the index, there are sub entries for file consistency, nonblocking communication, and point-to-point; there is a semantics and correctness entry for one-sided. Missing are collective, dynamic processes, datatypes, ...

**Comment by dholmes-epcc-ed-ac-uk** [comment]
@wesbland seems like this one should be assigned to all chapter chairs.

**Comment by wesbland** [comment]
Oops. You're right. I missed that.

**Comment by wesbland** [comment]
Each committee chair should check the box when they're done. Creating a small pull request for each chapter (or group of chapters if you own more than one) is probably as good a way as any to do this. Make sure that in your pull request, you mention this ticket (#237) without saying "Fixes" to avoid automatically closing it since there will be more than one PR to do this work. Also, if there's nothing to be done in your chapter, just check the box.

* [ ] Front matter - @wgropp 
* [ ] Introduction - @wgropp 
* [ ] Terms - @cblaas 
* [ ] Point-to-Point - @dholmes-epcc-ed-ac-uk 
* [ ] Datatypes - @bosilca 
* [ ] Collectives - @tonyskjellum 
* [ ] Communicators - @GuillaumeMercier 
* [ ] Topologies - @RolfRabenseifner 
* [ ] Partitioned Communication - @regrant 
* [ ] Environment - @bosilca 
* [ ] Info - @schulzm 
* [ ] Process Creation - @hppritcha 
* [ ] RMA - @wgropp 
* [ ] External Interfaces - @schulzm 
* [ ] I/O - @tonyskjellum 
* [ ] Tools - @mahermanns 
* [ ] Deprecated - @RolfRabenseifner 
* [ ] Removed - @RolfRabenseifner 
* [x] Backward Incompatibilities - @wesbland 
* [ ] Bindings - @puribangalore 
* [ ] Bindings Summary - @RolfRabenseifner 
* [ ] Change Log - @RolfRabenseifner

**Comment by wesbland** [comment]
Oh, I see now. This was going to be the next one we discussed next week. I was going to tag people after that. 😄 Oh well...

**Comment by schulzm** [comment]
In the EI chapter, we reference

Based on the above, this should get a \mpitermtitleindex{semantics!local}, but this really make sense? We don't have a local in the semantics index, but instead a local - should we tie this to there?

**Comment by wgropp** [comment]
I'd be more descriptive on the index entry - in this case, something like semantics|local for MPI\_CANCEL , since that's the context here.

**Comment by tonyskjellum** [comment]
I/O chapter has this implemented:
PR:  https://github.com/mpi-forum/mpi-standard/pull/246
PDF: [mpi40-report-ticket237-io-15jun20.pdf](https://github.com/mpi-forum/mpi-issues/files/4780816/mpi40-report-ticket237-io-15jun20.pdf)

**Comment by RolfRabenseifner** [comment]
I didn't find semamtics related sections in 
- Topology (except references to other semantics, e.g., collectives, persisntent),
- Deprecated, Removed, Language Bindings, and Change-Log

**Comment by wesbland** [comment]
For @dholmes-epcc-ed-ac-uk: https://github.com/mpi-forum/mpi-issues/issues/189

**Comment by wesbland** [comment]
For @tonyskjellum: https://github.com/mpi-forum/mpi-issues/issues/215

**Comment by mahermanns** [comment]
I think I need a bit of clarification on what I need to look for here:

Searching through the tools chapter on the look for *operation*, I found this
>\mpifunc{MPI\_T\_EVENT\_HANDLE\_SET\_INFO} updates the hints of the
event-registration handle associated with \mpiarg{event\_registration} using the
hints provided in \mpiarg{info}. This operation has no effect on previously
set or defaulted hints that are not specified by \mpiarg{info}.

Here to me, *operation* should actually be *procedure*, right? I think there are multiple occurrences in the tools chapter where operation is used but procedure is meant.

Here is an example where *procedure* is meant, yet the use of *operation* in this context is so common that it sounds weird to my ear to reword this to *receive procedures*. 
> have to be extended to have similar wrappers for all receive operations.

This would have to be *receive procedures* because I cannot wrap an operation, right?

**Comment by wesbland** [comment]
I think the intent from @wgropp here is for things that would go in the general index (i.e., concepts as opposed to specific function/constant references).

**Comment by dholmes-epcc-ed-ac-uk** [comment]
Does the "operation" have 4 stages: initialisation, starting, completion, and freeing? Even if there is currently only a single procedure in the standard that expresses a blocking operation (all four stages occur during one procedure call, before its return), is there anyway conceptually that these 4 stages could be split out into 4 separate procedures that fit with the semantics defined for them in chapter 2?

If `MPI_T_EVENT_HANDLE_SET_INFO` is collective, then it probably is a procedure that performs a blocking operation; we might want a nonblocking operation so that we can overlap compute with the network latency, and we might want a persistent variant to permit amortisation of setup costs for users that want to do this operation frequently.

I guess it is local, though - is this procedure (and its intended action) more like MPI_COMM_SET_INFO (collective -> probably an operation) or like MPI_COMM_SET_NAME (local -> probably not an operation).

**Comment by mahermanns** [comment]
@dholmes-epcc-ed-ac-uk All of `MPI_T` is local, so this in particular is not an operation.

**Comment by dholmes-epcc-ed-ac-uk** [comment]
I created PR https://github.com/mpi-forum/mpi-standard/pull/475 to deal with the point-to-point chapter portion of this issue.

The PR is much larger than I thought it would be when I started working on this. Given time constraints, it may be necessary to postpone addressing this issue until MPI-4.1---when we will have more time to review all the edits in it.

**Comment by cblaas** [comment]
I created PR https://github.com/mpi-forum/mpi-standard/pull/489 to deal with the Chapter 2 (terms) portion of this issue.

**Comment by GuillaumeMercier** [comment]
> The PR is much larger than I thought it would be when I started working on this. Given time constraints, it may be necessary to postpone addressing this issue until MPI-4.1---when we will have more time to review all the edits in it.

I addressed this a while back for Chapter 7 but did not a job as thorough as @dholmes-epcc-ed-ac-uk's. I definitively support the idea to get back to this issue for MPI  4.1.

**Comment by wesbland** [comment]
Closing this in favor of the chapter-specific issues.
